### PR TITLE
MAINT: Inline gentype_getreadbuf

### PR DESCRIPTION
**Pull request by sethtroisi**
This was proposed in one of the C cleanups, I tried it out and it looked better.

### CODE REVIEW

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 1]
```diff
@@ -2449,31 +2444,6 @@ static PySequenceMethods voidtype_as_sequence = {
...
-        PyErr_SetString(PyExc_SystemError,
-                "Accessing non-existent array segment");
-        return -1;
-    }
-
-    outcode = PyArray_DescrFromScalar(self);
-    numbytes = outcode->elsize;
-    *ptrptr = (void *)scalar_value(self, outcode);
-
-#ifndef Py_UNICODE_WIDE
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/scalartypes.c.src]: Is Py_UNICODE_WIDE no longer a thing? I think I'd prefer we tackle that ifdef in a separate patch, and keep this part for now.

Comment (eric-wieser) [reply numpy/core/src/multiarray/scalartypes.c.src]: That is, let's either:

* Inline the ifdef too, for now
* Put this patch on hold, and make a patch to remove `Py_UNICODE_WIDE` first

Comment (eric-wieser) [reply numpy/core/src/multiarray/scalartypes.c.src]: Just checked - `Py_UNICODE_WIDE` seems to be alive and well, although we'll stop needing when I fixup #15385

Comment (sethtroisi) [reply numpy/core/src/multiarray/scalartypes.c.src]: My reading of the code is that `Py_UNICODE_WIDE` only influences numbytes which is but used by the call site.

Comment (eric-wieser) [reply numpy/core/src/multiarray/scalartypes.c.src]: You're right, thanks for spotting that

**Review by eric-wieser — APPROVED** [review]

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 2]
```diff
@@ -1581,7 +1573,10 @@ gentype_byteswap(PyObject *self, PyObject *args, PyObject *kwds)
         PyObject *new;
         char *newmem;
 
-        gentype_getreadbuf(self, 0, (void **)&data);
+        PyArray_Descr *outcode = PyArray_DescrFromScalar(self);
+        data = (void *)scalar_value(self, outcode);
+        Py_DECREF(outcode);
+
         descr = PyArray_DescrFromScalar(self);
```
Comment (seberg) [inline numpy/core/src/multiarray/scalartypes.c.src]: This line is repeated now, maybe we can merge them

**Review by seberg — APPROVED** [review]
I am OK with just putting it in as is but if you have time, happy to reduce things a bit more even. I am curious if we can somehow remove that whole Malloc+free block, since we are handling a new object we are creating anyway, it is fine to byteswap in-place. But I am a bit worried of missing a corner case.

File: numpy/core/src/multiarray/scalartypes.c.src (hunk 2 above)
Comment (sethtroisi) [reply numpy/core/src/multiarray/scalartypes.c.src]: I'm less comfortable in python c code, and I consolidated two Py_DECREF to one place which felt better but made me question why they were seperated before. Can you give my update a through inspection.

Comment (seberg) [reply numpy/core/src/multiarray/scalartypes.c.src]: You missed that `descr` is still being used further down in `new = ...`

Comment (eric-wieser) [reply numpy/core/src/multiarray/scalartypes.c.src]: descr is a better name than outcode anyway

Comment (sethtroisi) [reply numpy/core/src/multiarray/scalartypes.c.src]: > You missed that `descr` is still being used further down in `new = ...`

I'm not sure what you're implying by this? I dropped `outcode` so I think desc should still be used in the `new = ...`.

Comment (seberg) [reply numpy/core/src/multiarray/scalartypes.c.src]: Yes, but as long as you are using `descr` you must hold a reference to it. If you call `DECREF` you lose that reference (i.e. if you wanted to be super clean, at least in this case, you might even do `descr = NULL` or similar incarnations to point out that it is not anymore valid to use).

Comment (sethtroisi) [reply numpy/core/src/multiarray/scalartypes.c.src]: Thanks, I was being silly and only making sure the count was right at the end of the function. Fixed now.

**Pull request merged**